storage-operator/pkg/controller/volume/volume_controller.go

```diff
@@ -24,149 +26,761 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
+const VOLUME_PROTECTION = "storage.metalk8s.scality.com/volume-protection"
+
 var log = logf.Log.WithName("controller_volume")
 
-// Add creates a new Volume Controller and adds it to the Manager. The Manager will set fields on the Controller
-// and Start it when the Manager is Started.
+// Creates a new Volume Controller and adds it to the Manager.
+//
+// The Manager will set fields on the Controller and Start it when the
+// Manager is Started.
 func Add(mgr manager.Manager) error {
 	return add(mgr, newReconciler(mgr))
 }
 
-// newReconciler returns a new reconcile.Reconciler
+// Returns a new reconcile.Reconciler.
 func newReconciler(mgr manager.Manager) reconcile.Reconciler {
 	return &ReconcileVolume{
-		client: mgr.GetClient(),
-		scheme: mgr.GetScheme(),
-		salt:   salt.NewClient(getAuthCredential(mgr.GetConfig())),
+		client:   mgr.GetClient(),
+		scheme:   mgr.GetScheme(),
+		recorder: mgr.GetRecorder("volume_controller"),
+		salt:     salt.NewClient(getAuthCredential(mgr.GetConfig())),
```

Comment:
SaltAPI auth tokens expire (I believe after 24h). What will happen in that case?

Reply:
Yup, lifetime is 12h by default.

It's handled in `authenticatedRequest`:

```go
	// Authenticate if we don't have a valid token.
	if self.token == nil || self.token.isExpired() {
		if err := self.authenticate(); err != nil {
			return nil, err
		}
	}

	response, err := self.doRequest(verb, endpoint, payload, true)
```

We re-auth when we detect the expiration (Salt returns an expiration date alongside the token).

Reply:
I also handled the 401 in case the token is invalidated before its expiration (can happen if the server restart, token are not persistent) => in such case we re-auth and retry once.